algorithms/kernel/svm/oneapi/cl_kernels/svm_train_block_smo_oneapi.cl

```diff
@@ -0,0 +1,232 @@
+    }
+
+    void reduceMax(const __local algorithmFPType * values, __local int * indices) {
+        const int local_id = get_local_id(0);
+        indices[local_id]  = local_id;
+
+        for (int stride = WS_SIZE / 2; stride > 0; stride >>= 1)
+        {
+            barrier(CLK_LOCAL_MEM_FENCE);
+
+            if (local_id < stride)
+            {
+                const algorithmFPType v  = values[indices[local_id]];
+                const algorithmFPType vk = values[indices[local_id + stride]];
+                if (vk > v)
+                {
+                    indices[local_id] = indices[local_id + stride];
+                }
+            }
+        }
+    }
+
+    void reduceMin(const __local algorithmFPType * values, __local int * indices) {
+        const int local_id = get_local_id(0);
+        indices[local_id]  = local_id;
+
+        for (int stride = WS_SIZE / 2; stride > 0; stride >>= 1)
+        {
+            barrier(CLK_LOCAL_MEM_FENCE);
+
+            if (local_id < stride)
+            {
+                const algorithmFPType v  = values[indices[local_id]];
+                const algorithmFPType vk = values[indices[local_id + stride]];
+                if (vk < v)
+                {
+                    indices[local_id] = indices[local_id + stride];
+                }
+            }
+        }
+    }
+
+    __kernel void smoKernel(const __global algorithmFPType * const y, const __global algorithmFPType * const kernelWsRows,
+                            const __global int * wsIndices, const uint ldx, const __global algorithmFPType * grad, const algorithmFPType C,
+                            const algorithmFPType eps, const algorithmFPType tau, const int maxInnerIteration, __global algorithmFPType * alpha,
+                            __global algorithmFPType * deltaalpha, __global algorithmFPType * resinfo) {
+        const uint i = get_local_id(0);
+        __local algorithmFPType kd[WS_SIZE];
+
+        const int wsIndex = wsIndices[i];
+
+        const algorithmFPType MIN_FLT = -FLT_MAX;
+        const algorithmFPType MAX_FLT = FLT_MAX;
+
+        const algorithmFPType two = 2.0;
+
+        algorithmFPType gradi           = grad[wsIndex];
+        algorithmFPType alphai          = alpha[wsIndex];
+        const algorithmFPType oldalphai = alphai;
+        const algorithmFPType yi        = y[wsIndex];
+
+        __local algorithmFPType objFunc[WS_SIZE];
+        __local int indices[WS_SIZE];
+
+        __local algorithmFPType deltaBi;
+        __local algorithmFPType deltaBj;
+
+        __local int Bi;
+        __local int Bj;
+
+        algorithmFPType ma;
+
+        kd[i] = kernelWsRows[i * ldx + wsIndex];
+        barrier(CLK_LOCAL_MEM_FENCE);
+
+        __local algorithmFPType localDiff;
+        __local algorithmFPType localEps;
+
+        int iter = 0;
+        for (; iter < maxInnerIteration; iter++)
+        {
+            /* m(alpha) = min(grad[i]): i belongs to I_UP (alpha) */
```

Comment:
Why? Do you mean to delete it?

Reply:
`alpha` - set argument, on which the decision of the function depends. this is not an index.  `I_UP (alpha)` - a set of available values. (see up). These comments are aligned with the comments of the CPU implementation of the SMO kernel. I left it so that the reader could find similar mathematical computation.